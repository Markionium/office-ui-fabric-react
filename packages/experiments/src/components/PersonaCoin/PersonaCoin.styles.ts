import { IPersonaCoinComponent } from './PersonaCoin.types';
import { initialsColorPropToColorCode } from 'office-ui-fabric-react/lib/components/Persona/PersonaInitialsColor';

export const DEFAULT_PERSONA_COIN_SIZE = 48;

export const PersonaCoinStyles: IPersonaCoinComponent['styles'] = props => {
  const { size = DEFAULT_PERSONA_COIN_SIZE, coinColor = initialsColorPropToColorCode(props), initialsColor = 'white' } = props;

  return {
    root: {
      position: 'relative',
      backgroundColor: coinColor,
      color: initialsColor,
      width: size,
      height: size,
      borderRadius: '50%',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center'
    },
    initials: {
      fontSize: `${size / 2.5}px`
    },
    image: {
      position: 'absolute',
      top: 0,
      left: 0,
      width: size,
      height: size,
      borderRadius: '50%',
      overflow: 'hidden'
    },
    presence: {
      position: 'absolute',
      // TODO: Pull out the following calculation into a function
      right: `-${size / 3 / 5}px`,
      bottom: `-${size / 3 / 5}px`
    }
  };
};
