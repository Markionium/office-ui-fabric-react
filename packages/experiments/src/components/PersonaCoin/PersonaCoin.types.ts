import { IStyle } from 'office-ui-fabric-react/lib/Styling';
import { IStatelessComponent, IStyleableComponentProps } from '../../Foundation';
import { ImageLoadState } from 'office-ui-fabric-react';

export type IPersonaCoinComponent = IStatelessComponent<IPersonaCoinProps, IPersonaCoinStyles>;

// Extending IStyleableComponentProps will automatically add stylable props for you, such as styles and theme.
//    If you don't want these props to be included in your component, just remove this extension.
export interface IPersonaCoinProps extends IStyleableComponentProps<IPersonaCoinProps, IPersonaCoinStyles> {
  initials?: string;
  /**
   * Size of the coin
   */
  size: 56 | 72 | 100;

  imageUrl?: string;

  color: string;

  onPhotoLoadingStateChange?: (loadState: ImageLoadState) => void;
}

export interface IPersonaCoinStyles {
  /**
   * Styling for the root element.
   */
  root: IStyle;
  /**
   * Styling for the image element
   */
  image: IStyle;
}
