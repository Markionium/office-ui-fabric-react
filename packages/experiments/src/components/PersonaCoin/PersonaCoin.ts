import { PersonaCoinView } from './PersonaCoin.view';
import { PersonaCoinStyles } from './PersonaCoin.styles';
import { PersonaCoinState } from './PersonaCoin.state';
import { IPersonaCoinProps, IPersonaCoinViewProps, IPersonaCoinStyles } from './PersonaCoin.types';
import { createComponent } from '../../Foundation';

export const PersonaCoin: React.StatelessComponent<IPersonaCoinProps> = createComponent<
  IPersonaCoinProps,
  IPersonaCoinViewProps,
  IPersonaCoinStyles
>({
  displayName: 'PersonaCoin',
  view: PersonaCoinView,
  styles: PersonaCoinStyles,
  state: PersonaCoinState
});
