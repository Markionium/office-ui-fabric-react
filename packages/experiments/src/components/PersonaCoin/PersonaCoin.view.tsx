/** @jsx withSlots */
import { withSlots, getSlots } from '../../Foundation';
import { PersonaPresence } from '../../utilities/factoryComponents';
import { IPersonaCoinComponent, IPersonaCoinProps, IPersonaCoinSlots } from './PersonaCoin.types';
import { PersonaCoinImage } from './PersonaCoinImage/PersonaCoinImage';
import { DEFAULT_PERSONA_COIN_SIZE } from './PersonaCoin.styles';
import { hideInitialsWhenImageIsLoaded } from './propHelpers';
import CoinAlternativeForSmallestSize from './CoinAlternativeForSmallestSize/CoinAlternativeForSmallestSize';
import { PersonaCoinInitials } from './PersonaCoinInitials/PersonaCoinInitials';

export const PersonaCoinView: IPersonaCoinComponent['view'] = props => {
  const coinSize = props.size || DEFAULT_PERSONA_COIN_SIZE;

  const Slots = getSlots<IPersonaCoinProps, IPersonaCoinSlots>(props, {
    root: 'div',
    image: PersonaCoinImage,
    initials: PersonaCoinInitials,
    presence: PersonaPresence,
    coinAlternativeForSmallestSize: CoinAlternativeForSmallestSize
  });

  if (coinSize === 10) {
    if (props.presence) {
      // TODO: why do we need to pass size twice?
      return <Slots.presence coinSize={coinSize} size={coinSize} />;
    }
    return <Slots.coinAlternativeForSmallestSize />;
  }

  const initials = hideInitialsWhenImageIsLoaded(props) ? null : (
    <Slots.initials initials={props.initials} text={props.text} allowPhoneInitials={props.allowPhoneInitials} />
  );

  return (
    <Slots.root>
      {initials}
      <Slots.image
        src={props.imageUrl}
        dimension={coinSize}
        onPhotoLoadingStateChange={props.onPhotoLoadingStateChange}
        imageShouldFadeIn={props.imageShouldFadeIn}
        imageShouldStartVisible={props.imageShouldStartVisible}
        imageAlt={props.imageAlt}
      />
      <Slots.presence coinSize={coinSize} size={coinSize} />
    </Slots.root>
  );
};
