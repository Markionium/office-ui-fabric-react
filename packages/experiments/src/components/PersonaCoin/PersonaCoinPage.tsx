import * as React from 'react';
import { ExampleCard, IComponentDemoPageProps, ComponentPage, PropertiesTableSet } from '@uifabric/example-app-base';

import { PersonaCoinExample } from './examples/PersonaCoin.Example';
const PersonaCoinExampleCode = require('!raw-loader!@uifabric/experiments/src/components/PersonaCoin/examples/PersonaCoin.Example.tsx') as string;

export class PersonaCoinPage extends React.Component<IComponentDemoPageProps, {}> {
  public render(): JSX.Element {
    return (
      <ComponentPage
        title=" PersonaCoin"
        componentName=" PersonaCoin"
        exampleCards={
          <div>
            <ExampleCard title=" PersonaCoin Ramps" code={PersonaCoinExampleCode}>
              <PersonaCoinExample />
            </ExampleCard>
          </div>
        }
        propertiesTables={
          <PropertiesTableSet
            sources={[require<string>('!raw-loader!@uifabric/experiments/src/components/PersonaCoin/PersonaCoin.types.ts')]}
          />
        }
        overview={<div />}
        bestPractices={<div />}
        dos={
          <div>
            <ul>
              <li />
            </ul>
          </div>
        }
        donts={
          <div>
            <ul>
              <li />
            </ul>
          </div>
        }
        isHeaderVisible={this.props.isHeaderVisible}
      />
    );
  }
}
