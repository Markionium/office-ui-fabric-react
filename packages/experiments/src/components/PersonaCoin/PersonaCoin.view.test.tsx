import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { IPersonaCoinStyles } from './PersonaCoin.types';
import { PersonaCoinView } from './PersonaCoin.view';
import { IProcessedStyleSet } from 'office-ui-fabric-react/lib/Styling';
import { PersonaTestImages } from '@uifabric/experiments/lib/common/TestImages';

// These tests will ensure that your styles regions have classname representation in snapshot output.
// (Also, classNames is a required prop for views, so we have to supply it for tests.)
const testPersonaCoinClassNames: IProcessedStyleSet<IPersonaCoinStyles> = {
  root: 'test-cn-root',
  image: 'test-cn-image',
  initials: 'test-cn-initials',
  presence: 'test-cn-presence',
  subComponentStyles: {}
};

// Views are just pure functions with no statefulness, which means they can get full code coverage
//    with snapshot tests exercising permutations of the props.
describe('PersonaCoinView', () => {
  it('renders a correct persona', () => {
    const tree = renderer.create(<PersonaCoinView text="James Bond" classNames={testPersonaCoinClassNames} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a coin with the initials JB', () => {
    const tree = renderer.create(<PersonaCoinView initials="JB" classNames={testPersonaCoinClassNames} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a red coin', () => {
    const tree = renderer.create(<PersonaCoinView initials="JB" classNames={testPersonaCoinClassNames} coinColor="red" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a coin with a contact icon', () => {
    const tree = renderer.create(<PersonaCoinView classNames={testPersonaCoinClassNames} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a coin with a contact icon for a Chinese name', () => {
    const tree = renderer.create(<PersonaCoinView text="五号" classNames={testPersonaCoinClassNames} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders presence when it is passed', () => {
    const tree = renderer.create(<PersonaCoinView text="五号" presence={4} classNames={testPersonaCoinClassNames} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders presence correctly when a very large coin is rendered', () => {
    const tree = renderer.create(<PersonaCoinView text="五号" presence={4} classNames={testPersonaCoinClassNames} size={100} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the coin with the provided image', () => {
    const tree = renderer
      .create(<PersonaCoinView text="Ellen Grace" imageUrl={PersonaTestImages.personMale} classNames={testPersonaCoinClassNames} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
