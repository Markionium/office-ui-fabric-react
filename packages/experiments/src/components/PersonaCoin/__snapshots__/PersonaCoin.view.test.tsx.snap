// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonaCoinView renders a coin with a contact icon 1`] = `
<div
  className="test-cn-root"
>
  <i
    aria-hidden={true}
    className=

        {
          display: inline-block;
        }
    data-icon-name="Contact"
    role="presentation"
  />
</div>
`;

exports[`PersonaCoinView renders a coin with a contact icon for a Chinese name 1`] = `
<div
  className="test-cn-root"
>
  <i
    aria-hidden={true}
    className=

        {
          display: inline-block;
        }
    data-icon-name="Contact"
    role="presentation"
  />
</div>
`;

exports[`PersonaCoinView renders a coin with the initials JB 1`] = `
<div
  className="test-cn-root"
>
  <span
    className=
        test-cn-initials
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          color: inherit;
          display: block;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: inherit;
          font-weight: inherit;
          overflow: hidden;
          text-overflow: ellipsis;
          white-space: nowrap;
        }
  >
    JB
  </span>
</div>
`;

exports[`PersonaCoinView renders a correct persona 1`] = `
<div
  className="test-cn-root"
>
  <span
    className=
        test-cn-initials
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          color: inherit;
          display: block;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: inherit;
          font-weight: inherit;
          overflow: hidden;
          text-overflow: ellipsis;
          white-space: nowrap;
        }
  >
    JB
  </span>
</div>
`;

exports[`PersonaCoinView renders a red coin 1`] = `
<div
  className="test-cn-root"
>
  <span
    className=
        test-cn-initials
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          color: inherit;
          display: block;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: inherit;
          font-weight: inherit;
          overflow: hidden;
          text-overflow: ellipsis;
          white-space: nowrap;
        }
  >
    JB
  </span>
</div>
`;

exports[`PersonaCoinView renders presence correctly when a very large coin is rendered 1`] = `
<div
  className="test-cn-root"
>
  <i
    aria-hidden={true}
    className=

        {
          display: inline-block;
        }
    data-icon-name="Contact"
    role="presentation"
  />
  <div
    className=
        ms-Persona-presence
        test-cn-presence
        {
          -ms-high-contrast-adjust: none;
          background-clip: content-box;
          background-color: #E81123;
          border-radius: 50%;
          border: 2px solid #ffffff;
          bottom: -2px;
          box-sizing: content-box;
          height: 12px;
          position: absolute;
          right: -2px;
          text-align: center;
          top: auto;
          width: 12px;
        }
        @media screen and (-ms-high-contrast: active){& {
          background-color: WindowText;
          border-color: Window;
        }
    style={
      Object {
        "height": "33.333333333333336px",
        "width": "33.333333333333336px",
      }
    }
  >
    <i
      aria-hidden={true}
      className=
          ms-Persona-presenceIcon
          {
            color: #ffffff;
            display: inline-block;
            font-size: 6px;
            line-height: 12px;
            vertical-align: top;
          }
          @media screen and (-ms-high-contrast: active){& {
            color: Window;
          }
      data-icon-name="SkypeMinus"
      role="presentation"
      style={
        Object {
          "fontSize": "16.666666666666668px",
          "lineHeight": "33.333333333333336px",
        }
      }
    />
  </div>
</div>
`;

exports[`PersonaCoinView renders presence when it is passed 1`] = `
<div
  className="test-cn-root"
>
  <i
    aria-hidden={true}
    className=

        {
          display: inline-block;
        }
    data-icon-name="Contact"
    role="presentation"
  />
  <div
    className=
        ms-Persona-presence
        test-cn-presence
        {
          -ms-high-contrast-adjust: none;
          background-clip: content-box;
          background-color: #E81123;
          border-radius: 50%;
          border: 2px solid #ffffff;
          bottom: -2px;
          box-sizing: content-box;
          height: 12px;
          position: absolute;
          right: -2px;
          text-align: center;
          top: auto;
          width: 12px;
        }
        @media screen and (-ms-high-contrast: active){& {
          background-color: WindowText;
          border-color: Window;
        }
    style={
      Object {
        "height": "16px",
        "width": "16px",
      }
    }
  >
    <i
      aria-hidden={true}
      className=
          ms-Persona-presenceIcon
          {
            color: #ffffff;
            display: inline-block;
            font-size: 6px;
            line-height: 12px;
            vertical-align: top;
          }
          @media screen and (-ms-high-contrast: active){& {
            color: Window;
          }
      data-icon-name="SkypeMinus"
      role="presentation"
      style={
        Object {
          "fontSize": "8px",
          "lineHeight": "16px",
        }
      }
    />
  </div>
</div>
`;

exports[`PersonaCoinView renders the coin with the provided image 1`] = `
<div
  className="test-cn-root"
>
  <span
    className=
        test-cn-initials
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          color: inherit;
          display: block;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: inherit;
          font-weight: inherit;
          overflow: hidden;
          text-overflow: ellipsis;
          white-space: nowrap;
        }
  >
    EG
  </span>
  <div
    className=
        ms-Image
        test-cn-image
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          border-radius: 50%;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 400;
          overflow: hidden;
          position: relative;
        }
    style={
      Object {
        "height": 48,
        "width": 48,
      }
    }
  >
    <img
      className=
          ms-Image-image
          ms-Image-image--cover
          ms-Image-image--portrait
          is-notLoaded
          is-fadeIn
          {
            display: block;
            height: auto;
            left: 50% /* @noflip */;
            opacity: 0;
            position: absolute;
            top: 50%;
            transform: translate(-50%,-50%);
            width: 100%;
          }
      onError={[Function]}
      onLoad={[Function]}
      src="https://static2.sharepointonline.com/files/fabric/office-ui-fabric-react-assets//persona-male.png"
    />
  </div>
</div>
`;
